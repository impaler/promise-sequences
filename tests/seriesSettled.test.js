import test from 'ava'
const {seriesSettled} = require('../src/promise-sequences')
const {rejectTimeout, resolveTimeout, methodThatThrows, methodThatReturns} = require('./helpers')

test(`seriesSettled default settings will resolve
      three promises one at a time`, async t => {
  const tasks = [
    resolveTimeout(1),
    resolveTimeout('test'),
    resolveTimeout(2),
  ]
  const results = await seriesSettled(tasks)

  t.snapshot(results)
})

test(`should execute synchronous functions and catch the settled
      results, not throwing errors`, async t => {
  const tasks = [
    methodThatThrows,
    methodThatReturns,
    methodThatThrows,
  ]
  const results = await seriesSettled(tasks)

  t.snapshot(results)
})

test(`should execute promises, functions or values together
      whether they reject or resolve`, async t => {
  const tasks = [
    'first value',
    Promise.reject('second value'),
    Promise.resolve(3),
    resolveTimeout(4),
    rejectTimeout(5),
    resolveTimeout.bind(null, 6),
    rejectTimeout.bind(null, 7),
  ]
  const results = await seriesSettled(tasks)

  t.snapshot(results)
})

test(`should execute promises in series, any promise that rejects
      will not reject the entire series.`, async t => {
  const tasks = [
    rejectTimeout(1),
    rejectTimeout(2),
    resolveTimeout(3),
    rejectTimeout(4),
    rejectTimeout(5),
  ]
  const results = await seriesSettled(tasks)

  t.snapshot(results)
})

test(`should execute a series of promises when the concurrency is
      larger than the count of the promises`, async t => {
  const concurrency = 4
  const tasks = [
    1,
    resolveTimeout(2),
    resolveTimeout.bind(null, 3),
  ]
  const results = await seriesSettled(tasks, concurrency)

  t.snapshot(results)
})

test(`execute a series of promises when the count of the promises
      is the same as the concurrency`, async t => {
  const concurrency = 3
  const tasks = [
    Promise.reject(1),
    rejectTimeout(2),
    rejectTimeout.bind(null, 3),
  ]
  const results = await seriesSettled(tasks, concurrency)

  t.snapshot(results)
})

test(`should execute execute a series with a concurrency limit of 2.
      The step function will output context on each concurrent step.
      Any promise that rejects will not reject the entire series.`, async t => {
  const concurrent = 2
  const tasks = [
    Promise.resolve(1),
    Promise.resolve(2),
    resolveTimeout(3),
    rejectTimeout(4),
    Promise.resolve(5),
    resolveTimeout(6),
  ]
  const steps = []
  const step = (value, current, total) => steps.push({value, current, total})
  const results = await seriesSettled(tasks, concurrent, step)

  t.snapshot(steps)
  t.snapshot(results)
})

test(`should should execute promises with no concurrency limit,
      any promise that rejects will not reject the entire series.`, async t => {
  const tasks = [
    resolveTimeout(1),
    resolveTimeout(2),
    rejectTimeout(3),
    resolveTimeout(4),
    rejectTimeout(5),
  ]
  const steps = []
  const step = (value, current, total) => steps.push({value, current, total})
  const results = await seriesSettled(tasks, 3, step)

  t.snapshot(steps)
  t.snapshot(results)
})
