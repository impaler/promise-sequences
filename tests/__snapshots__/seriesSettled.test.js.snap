exports[`execute a series of promises when the count of the promises
      is the same as the concurrency 1`] = `"[{\"state\":\"rejected\",\"result\":1},{\"state\":\"rejected\",\"result\":2},{\"state\":\"rejected\",\"result\":3}]"`;

exports[`seriesSettled default settings will resolve
      three promises one at a time 1`] = `"[{\"state\":\"resolved\",\"result\":1},{\"state\":\"resolved\",\"result\":\"test\"},{\"state\":\"resolved\",\"result\":2}]"`;

exports[`should execute a series of promises when the concurrency is
      larger than the count of the promises 1`] = `"[{\"state\":\"resolved\",\"result\":1},{\"state\":\"resolved\",\"result\":2},{\"state\":\"resolved\",\"result\":3}]"`;

exports[`should execute execute a series with a concurrency limit of 2.
      The step function will output context on each concurrent step.
      Any promise that rejects will not reject the entire series. 1`] = `"[{\"value\":[{\"state\":\"resolved\",\"result\":1},{\"state\":\"resolved\",\"result\":2}],\"current\":1,\"total\":3},{\"value\":[{\"state\":\"resolved\",\"result\":1},{\"state\":\"resolved\",\"result\":2},{\"state\":\"resolved\",\"result\":3},{\"state\":\"rejected\",\"result\":4}],\"current\":2,\"total\":3},{\"value\":[{\"state\":\"resolved\",\"result\":1},{\"state\":\"resolved\",\"result\":2},{\"state\":\"resolved\",\"result\":3},{\"state\":\"rejected\",\"result\":4},{\"state\":\"resolved\",\"result\":5},{\"state\":\"resolved\",\"result\":6}],\"current\":3,\"total\":3}]"`;

exports[`should execute execute a series with a concurrency limit of 2.
      The step function will output context on each concurrent step.
      Any promise that rejects will not reject the entire series. 2`] = `"[{\"state\":\"resolved\",\"result\":1},{\"state\":\"resolved\",\"result\":2},{\"state\":\"resolved\",\"result\":3},{\"state\":\"rejected\",\"result\":4},{\"state\":\"resolved\",\"result\":5},{\"state\":\"resolved\",\"result\":6}]"`;

exports[`should execute promises in series, any promise that rejects
      will not reject the entire series. 1`] = `"[{\"state\":\"rejected\",\"result\":1},{\"state\":\"rejected\",\"result\":2},{\"state\":\"resolved\",\"result\":3},{\"state\":\"rejected\",\"result\":4},{\"state\":\"rejected\",\"result\":5}]"`;

exports[`should execute promises, functions or values together
      whether they reject or resolve 1`] = `"[{"state":"resolved","result":"first value"},{"state":"rejected","result":"second value"},{"state":"resolved","result":3},{"state":"resolved","result":4},{"state":"rejected","result":5},{"state":"resolved","result":6},{"state":"rejected","result":7}]"`;

exports[`should execute synchronous functions and catch the settled
      results, not throwing errors 1`] = `"[{\"state\":\"rejected\",\"result\":\"bs\"},{\"state\":\"resolved\",\"result\":\"reason\"},{\"state\":\"rejected\",\"result\":\"bs\"}]"`;

exports[`should should execute promises with no concurrency limit,
      any promise that rejects will not reject the entire series. 1`] = `"[{\"value\":[{\"state\":\"resolved\",\"result\":1},{\"state\":\"resolved\",\"result\":2},{\"state\":\"rejected\",\"result\":3}],\"current\":1,\"total\":2},{\"value\":[{\"state\":\"resolved\",\"result\":1},{\"state\":\"resolved\",\"result\":2},{\"state\":\"rejected\",\"result\":3},{\"state\":\"resolved\",\"result\":4},{\"state\":\"rejected\",\"result\":5}],\"current\":2,\"total\":2}]"`;

exports[`should should execute promises with no concurrency limit,
      any promise that rejects will not reject the entire series. 2`] = `"[{\"state\":\"resolved\",\"result\":1},{\"state\":\"resolved\",\"result\":2},{\"state\":\"rejected\",\"result\":3},{\"state\":\"resolved\",\"result\":4},{\"state\":\"rejected\",\"result\":5}]"`;
